// @ironfish/sdk/build/src/mining/stratum/clients/tlsClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StratumTlsClient = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const tls_1 = __importDefault(require("tls"));
const tcpClient_1 = require("./tcpClient");
class StratumTlsClient extends tcpClient_1.StratumTcpClient {
    connect() {
        return new Promise((resolve, reject) => {
            const onConnect = () => {
                client.off('secureConnect', onConnect);
                client.off('error', onError);
                client.on('error', this.onError);
                client.on('close', this.onSocketDisconnect);
                resolve();
            };
            const onError = (error) => {
                client.off('secureConnect', onConnect);
                client.off('error', onError);
                reject(error);
            };
            const client = tls_1.default.connect({
                host: this.host,
                port: this.port,
                rejectUnauthorized: false,
            });
            client.on('error', onError);
            client.on('secureConnect', onConnect);
            client.on('data', this.onSocketData);
            this.client = client;
        });
    }
}
exports.StratumTlsClient = StratumTlsClient;
