// @ironfish/sdk/build/src/mining/stratum/adapters/tcpAdapter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StratumTcpAdapter = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const net_1 = __importDefault(require("net"));
class StratumTcpAdapter {
    constructor(options) {
        this.server = null;
        this.stratumServer = null;
        this.started = false;
        this.logger = options.logger;
        this.host = options.host;
        this.port = options.port;
    }
    createServer() {
        this.logger.info(`Hosting Stratum via TCP on ${this.host}:${this.port}`);
        return net_1.default.createServer((socket) => this.stratumServer?.onConnection(socket));
    }
    start() {
        if (this.started) {
            return Promise.resolve();
        }
        this.started = true;
        return new Promise((resolve, reject) => {
            try {
                this.server = this.createServer();
                this.server.listen(this.port, this.host, () => {
                    resolve();
                });
            }
            catch (e) {
                reject(e);
            }
        });
    }
    stop() {
        if (!this.started) {
            return Promise.resolve();
        }
        return new Promise((resolve, reject) => {
            this.server?.close((e) => {
                return e ? reject(e) : resolve();
            });
        });
    }
    attach(server) {
        this.stratumServer = server;
    }
}
exports.StratumTcpAdapter = StratumTcpAdapter;
