// @ironfish/sdk/build/src/rpc/routes/event/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RpcBlockHeaderSchema = exports.serializeRpcBlockHeader = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const yup = __importStar(require("yup"));
function serializeRpcBlockHeader(header) {
    return {
        hash: header.hash.toString('hex'),
        sequence: Number(header.sequence),
        previousBlockHash: header.previousBlockHash.toString('hex'),
        timestamp: header.timestamp.valueOf(),
        difficulty: header.target.toDifficulty().toString(),
        graffiti: header.graffiti.toString('hex'),
    };
}
exports.serializeRpcBlockHeader = serializeRpcBlockHeader;
exports.RpcBlockHeaderSchema = yup
    .object({
    hash: yup.string().defined(),
    sequence: yup.number().defined(),
    previousBlockHash: yup.string().defined(),
    timestamp: yup.number().defined(),
    transactions: yup.array().defined(),
    difficulty: yup.string().defined(),
    graffiti: yup.string().defined(),
})
    .defined();
