// @ironfish/sdk/build/src/rpc/routes/chain/broadcastTransaction.js
"use strict";
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BroadcastTransactionResponseSchema = exports.BroadcastTransactionRequestSchema = void 0;
const yup = __importStar(require("yup"));
const assert_1 = require("../../../assert");
const node_1 = require("../../../node");
const primitives_1 = require("../../../primitives");
const adapters_1 = require("../../adapters");
const router_1 = require("../router");
exports.BroadcastTransactionRequestSchema = yup
    .object({
    transaction: yup.string().defined(),
})
    .defined();
exports.BroadcastTransactionResponseSchema = yup
    .object({
    hash: yup.string().defined(),
    accepted: yup.boolean().defined(),
    broadcasted: yup.boolean().defined(),
})
    .defined();
router_1.routes.register(`${router_1.ApiNamespace.chain}/broadcastTransaction`, exports.BroadcastTransactionRequestSchema, async (request, node) => {
    assert_1.Assert.isInstanceOf(node, node_1.FullNode);
    const data = Buffer.from(request.data.transaction, 'hex');
    const transaction = new primitives_1.Transaction(data);
    const verify = await node.chain.verifier.verifyNewTransaction(transaction);
    if (!verify.valid) {
        throw new adapters_1.ValidationError(`Invalid transaction, reason: ${String(verify.reason)}`);
    }
    const accepted = node.memPool.acceptTransaction(transaction);
    let broadcasted = false;
    if (node.peerNetwork.isReady) {
        node.peerNetwork.broadcastTransaction(transaction);
        broadcasted = true;
    }
    request.end({
        accepted,
        broadcasted,
        hash: transaction.hash().toString('hex'),
    });
});
