// @ironfish/sdk/build/src/rpc/routes/chain/getConsensusParameters.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetConsensusParametersResponseSchema = exports.GetConsensusParametersRequestSchema = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const yup = __importStar(require("yup"));
const assert_1 = require("../../../assert");
const node_1 = require("../../../node");
const router_1 = require("../router");
exports.GetConsensusParametersRequestSchema = yup.mixed().oneOf([undefined]);
exports.GetConsensusParametersResponseSchema = yup
    .object({
    allowedBlockFuturesSeconds: yup.number().defined(),
    genesisSupplyInIron: yup.number().defined(),
    targetBlockTimeInSeconds: yup.number().defined(),
    targetBucketTimeInSeconds: yup.number().defined(),
    maxBlockSizeBytes: yup.number().defined(),
    minFee: yup.number().defined(),
})
    .defined();
router_1.routes.register(`${router_1.ApiNamespace.chain}/getConsensusParameters`, exports.GetConsensusParametersRequestSchema, (request, node) => {
    assert_1.Assert.isInstanceOf(node, node_1.FullNode);
    assert_1.Assert.isNotNull(node.chain.consensus, 'no consensus parameters');
    const consensusParameters = node.chain.consensus.parameters;
    request.end({
        allowedBlockFuturesSeconds: consensusParameters.allowedBlockFutureSeconds,
        genesisSupplyInIron: consensusParameters.genesisSupplyInIron,
        targetBlockTimeInSeconds: consensusParameters.targetBlockTimeInSeconds,
        targetBucketTimeInSeconds: consensusParameters.targetBucketTimeInSeconds,
        maxBlockSizeBytes: consensusParameters.maxBlockSizeBytes,
        minFee: consensusParameters.minFee,
    });
});
