// @ironfish/sdk/build/src/rpc/routes/miner/blockTemplateStream.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockTemplateStreamResponseSchema = exports.BlockTemplateStreamRequestSchema = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const yup = __importStar(require("yup"));
const assert_1 = require("../../../assert");
const node_1 = require("../../../node");
const router_1 = require("../router");
exports.BlockTemplateStreamRequestSchema = yup
    .mixed()
    .oneOf([undefined]);
exports.BlockTemplateStreamResponseSchema = yup
    .object({
    header: yup
        .object({
        sequence: yup.number().required(),
        previousBlockHash: yup.string().required(),
        noteCommitment: yup.string().required(),
        transactionCommitment: yup.string().required(),
        target: yup.string().required(),
        randomness: yup.string().required(),
        timestamp: yup.number().required(),
        graffiti: yup.string().required(),
    })
        .required()
        .defined(),
    transactions: yup.array().of(yup.string().required()).required().defined(),
    previousBlockInfo: yup
        .object({
        target: yup.string().required(),
        timestamp: yup.number().required(),
    })
        .required()
        .defined(),
})
    .required()
    .defined();
router_1.routes.register(`${router_1.ApiNamespace.miner}/blockTemplateStream`, exports.BlockTemplateStreamRequestSchema, (request, node) => {
    assert_1.Assert.isInstanceOf(node, node_1.FullNode);
    if (!node.chain.synced && !node.config.get('miningForce')) {
        node.logger.info('Miner connected while the node is syncing. Will not start mining until the node is synced');
    }
    const streamBlockTemplate = (serializedBlock) => {
        request.stream(serializedBlock);
    };
    node.miningManager.onNewBlockTemplate(streamBlockTemplate);
    // If the listener stops listening, we no longer need to generate new block templates
    request.onClose.once(() => {
        node.miningManager.offNewBlockTemplate(streamBlockTemplate);
    });
});
