// @ironfish/sdk/build/src/workerPool/tasks/jobAbort.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobAbortedError = exports.JobAbortedMessage = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const workerMessage_1 = require("./workerMessage");
class JobAbortedMessage extends workerMessage_1.WorkerMessage {
    constructor(jobId) {
        super(workerMessage_1.WorkerMessageType.JobAborted, jobId);
    }
    serializePayload() {
        return;
    }
    static deserializePayload() {
        return new JobAbortedError();
    }
    getSize() {
        return 0;
    }
}
exports.JobAbortedMessage = JobAbortedMessage;
class JobAbortedError extends Error {
    constructor() {
        super();
        this.name = this.constructor.name;
        this.type = 'JobAbortedError';
        this.name = 'JobAbortedError';
    }
}
exports.JobAbortedError = JobAbortedError;
