// @ironfish/sdk/build/src/network/messages/getCompactBlock.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetCompactBlockResponse = exports.GetCompactBlockRequest = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const bufio_1 = __importDefault(require("bufio"));
const types_1 = require("../types");
const serializers_1 = require("../utils/serializers");
const rpcNetworkMessage_1 = require("./rpcNetworkMessage");
class GetCompactBlockRequest extends rpcNetworkMessage_1.RpcNetworkMessage {
    constructor(blockHash, rpcId) {
        super(types_1.NetworkMessageType.GetCompactBlockRequest, rpcNetworkMessage_1.Direction.Request, rpcId);
        this.blockHash = blockHash;
    }
    serializePayload(bw) {
        bw.writeHash(this.blockHash);
    }
    static deserializePayload(buffer, rpcId) {
        const reader = bufio_1.default.read(buffer, true);
        const blockHash = reader.readHash();
        return new GetCompactBlockRequest(blockHash, rpcId);
    }
    getSize() {
        return 32;
    }
}
exports.GetCompactBlockRequest = GetCompactBlockRequest;
class GetCompactBlockResponse extends rpcNetworkMessage_1.RpcNetworkMessage {
    constructor(compactBlock, rpcId) {
        super(types_1.NetworkMessageType.GetCompactBlockResponse, rpcNetworkMessage_1.Direction.Response, rpcId);
        this.compactBlock = compactBlock;
    }
    serializePayload(bw) {
        serializers_1.writeCompactBlock(bw, this.compactBlock);
    }
    static deserializePayload(buffer, rpcId) {
        const reader = bufio_1.default.read(buffer, true);
        const compactBlock = serializers_1.readCompactBlock(reader);
        return new GetCompactBlockResponse(compactBlock, rpcId);
    }
    getSize() {
        return serializers_1.getCompactBlockSize(this.compactBlock);
    }
}
exports.GetCompactBlockResponse = GetCompactBlockResponse;
