// @ironfish/sdk/build/src/network/messages/rpcNetworkMessage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RpcNetworkMessage = exports.Direction = exports.RPC_TIMEOUT_MILLIS = void 0;
const bufio_1 = __importDefault(require("bufio"));
const networkMessage_1 = require("./networkMessage");
exports.RPC_TIMEOUT_MILLIS = 30000;
var Direction;
(function (Direction) {
    Direction["Request"] = "request";
    Direction["Response"] = "response";
})(Direction = exports.Direction || (exports.Direction = {}));
class RpcNetworkMessage extends networkMessage_1.NetworkMessage {
    constructor(type, direction, rpcId) {
        super(type);
        this.direction = direction;
        this.rpcId = rpcId ?? (RpcNetworkMessage.id = ++RpcNetworkMessage.id % 0xffff);
    }
    static deserializeHeader(buffer) {
        const br = bufio_1.default.read(buffer, true);
        const rpcId = br.readU16();
        const remaining = br.readBytes(br.left());
        return { rpcId, remaining };
    }
    serialize() {
        const headerSize = 3;
        const bw = bufio_1.default.write(headerSize + this.getSize());
        bw.writeU8(this.type);
        bw.writeU16(this.rpcId);
        this.serializePayload(bw);
        return bw.render();
    }
}
exports.RpcNetworkMessage = RpcNetworkMessage;
RpcNetworkMessage.id = 0;
