// @ironfish/sdk/build/src/fileStores/hosts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HostsStore = exports.HOST_FILE_NAME = exports.HostOptionsDefaults = void 0;
const logger_1 = require("../logger");
const json_1 = require("../utils/json");
const keyStore_1 = require("./keyStore");
exports.HostOptionsDefaults = {
    priorPeers: [],
};
exports.HOST_FILE_NAME = 'hosts.json';
class HostsStore extends keyStore_1.KeyStore {
    constructor(files, dataDir) {
        super(files, exports.HOST_FILE_NAME, exports.HostOptionsDefaults, dataDir);
        this.logger = logger_1.createRootLogger();
    }
    async load() {
        try {
            await super.load();
        }
        catch (e) {
            if (e instanceof json_1.ParseJsonError) {
                this.logger.debug(`Error: Could not parse JSON at ${this.storage.configPath}, overwriting file.`);
                await super.save();
            }
            else {
                throw e;
            }
        }
    }
}
exports.HostsStore = HostsStore;
