// @ironfish/sdk/build/src/utils/buffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BufferUtils = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const string_1 = require("./string");
function toHuman(buffer) {
    return string_1.StringUtils.sanitizeString(buffer.toString('utf8'));
}
function incrementLE(buffer) {
    for (let i = 0; i < buffer.length; i++) {
        if (buffer[i]++ !== 255) {
            break;
        }
    }
}
function incrementBE(buffer) {
    for (let i = buffer.length - 1; i >= 0; i--) {
        if (buffer[i]++ !== 255) {
            break;
        }
    }
}
function equalsNullable(a, b) {
    return a == null || b == null ? a === b : a.equals(b);
}
function maxNullable(a, b) {
    if (!a) {
        return b ? b : undefined;
    }
    else if (!b) {
        return a;
    }
    else {
        return Buffer.compare(a, b) > 0 ? a : b;
    }
}
function minNullable(a, b) {
    if (!a) {
        return b ? b : undefined;
    }
    else if (!b) {
        return a;
    }
    else {
        return Buffer.compare(a, b) <= 0 ? a : b;
    }
}
exports.BufferUtils = {
    toHuman,
    equalsNullable,
    incrementLE,
    incrementBE,
    maxNullable,
    minNullable,
};
