// @ironfish/sdk/build/src/mining/stratum/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerMessageMalformedError = exports.ClientMessageMalformedError = exports.MessageMalformedError = void 0;
class MessageMalformedError extends Error {
    constructor(sender, error, method) {
        super();
        this.name = this.constructor.name;
        if (typeof error === 'string') {
            this.message = error;
        }
        else {
            this.message = `${sender} sent malformed request`;
            if (method) {
                this.message += ` (${method})`;
            }
            this.message += `: ${error.message}`;
        }
    }
}
exports.MessageMalformedError = MessageMalformedError;
class ClientMessageMalformedError extends MessageMalformedError {
    constructor(client, error, method) {
        super(`Client ${client.id}`, error, method);
        this.client = client;
    }
}
exports.ClientMessageMalformedError = ClientMessageMalformedError;
class ServerMessageMalformedError extends MessageMalformedError {
    constructor(error, method) {
        super('Server', error, method);
    }
}
exports.ServerMessageMalformedError = ServerMessageMalformedError;
