// @ironfish/sdk/build/src/rpc/messageBuffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageBuffer = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const protocol_1 = require("./adapters/socketAdapter/protocol");
class MessageBuffer {
    constructor(delimiter = protocol_1.MESSAGE_DELIMITER) {
        this.delimiter = delimiter;
        this.buffer = '';
    }
    write(data) {
        this.buffer += data.toString('utf-8');
    }
    clear() {
        this.buffer = '';
    }
    readMessages() {
        const lastDelimiterIndex = this.buffer.lastIndexOf(this.delimiter);
        // buffer contains no full messages
        if (lastDelimiterIndex === -1) {
            return [];
        }
        const messages = this.buffer.substring(0, lastDelimiterIndex).trim().split(this.delimiter);
        this.buffer = this.buffer.substring(lastDelimiterIndex + 1);
        return messages;
    }
}
exports.MessageBuffer = MessageBuffer;
