// @ironfish/sdk/build/src/rpc/request.js
"use strict";
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RpcRequest = void 0;
const event_1 = require("../event");
class RpcRequest {
    constructor(data, route, onEnd, onStream) {
        this.ended = false;
        this.closed = false;
        this.onClose = new event_1.Event();
        this.data = data;
        this.route = route;
        this.onEnd = onEnd;
        this.onStream = onStream;
    }
    end(data, status) {
        if (this.ended) {
            throw new Error(`Request has already ended`);
        }
        this.ended = true;
        if (this.closed) {
            return;
        }
        this.onClose.clear();
        this.onEnd(status || 200, data);
    }
    stream(data) {
        if (this.closed) {
            return;
        }
        if (this.ended) {
            throw new Error(`Request has already ended`);
        }
        this.onStream(data);
    }
    close() {
        this.closed = true;
        this.onClose.emit();
    }
}
exports.RpcRequest = RpcRequest;
