// @ironfish/sdk/build/src/wallet/account/encoder/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonEncoder = void 0;
const validator_1 = require("../../validator");
const encoder_1 = require("./encoder");
class JsonEncoder {
    encode(value) {
        let createdAt = null;
        if (value.createdAt) {
            createdAt = {
                hash: value.createdAt.hash.toString('hex'),
                sequence: value.createdAt.sequence,
            };
        }
        return JSON.stringify({ ...value, createdAt });
    }
    decode(value, options) {
        let account;
        try {
            account = JSON.parse(value);
            if (account.createdAt && !account.createdAt.hash) {
                account.createdAt = null;
            }
            const accountImport = {
                ...account,
                name: options?.name ? options.name : account.name,
                createdAt: account.createdAt
                    ? {
                        hash: Buffer.from(account.createdAt.hash, 'hex'),
                        sequence: account.createdAt.sequence,
                    }
                    : null,
            };
            validator_1.validateAccount(accountImport);
            return accountImport;
        }
        catch (e) {
            throw new encoder_1.DecodeFailed(`Invalid JSON: ${e.message}`, this.constructor.name);
        }
    }
}
exports.JsonEncoder = JsonEncoder;
