// @ironfish/sdk/build/src/wallet/account/encoder/spendingKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpendingKeyEncoder = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const rust_nodejs_1 = require("@ironfish/rust-nodejs");
const account_1 = require("../account");
const encoder_1 = require("./encoder");
class SpendingKeyEncoder {
    encode(value) {
        if (!value.spendingKey) {
            throw new Error('Spending key is required for spending key encoder');
        }
        return value.spendingKey;
    }
    decode(spendingKey, options) {
        let key;
        try {
            key = rust_nodejs_1.generateKeyFromPrivateKey(spendingKey);
        }
        catch (e) {
            throw new encoder_1.DecodeFailed(`Invalid spending key: ${e.message}`, this.constructor.name);
        }
        if (!options.name) {
            throw new encoder_1.DecodeInvalid('Name option is required for spending key encoder');
        }
        return {
            name: options.name,
            spendingKey: spendingKey,
            viewKey: key.viewKey,
            incomingViewKey: key.incomingViewKey,
            outgoingViewKey: key.outgoingViewKey,
            publicAddress: key.publicAddress,
            createdAt: null,
            version: account_1.ACCOUNT_SCHEMA_VERSION,
        };
    }
}
exports.SpendingKeyEncoder = SpendingKeyEncoder;
