// @ironfish/sdk/build/src/network/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkMessageType = void 0;
var NetworkMessageType;
(function (NetworkMessageType) {
    // Core protocol messages
    NetworkMessageType[NetworkMessageType["Identify"] = 0] = "Identify";
    NetworkMessageType[NetworkMessageType["Disconnecting"] = 1] = "Disconnecting";
    NetworkMessageType[NetworkMessageType["SignalRequest"] = 2] = "SignalRequest";
    NetworkMessageType[NetworkMessageType["Signal"] = 3] = "Signal";
    NetworkMessageType[NetworkMessageType["CannotSatisfyRequest"] = 4] = "CannotSatisfyRequest";
    // Peer discovery-related messages
    NetworkMessageType[NetworkMessageType["PeerListRequest"] = 20] = "PeerListRequest";
    NetworkMessageType[NetworkMessageType["PeerList"] = 21] = "PeerList";
    // Syncing-related messages
    // Block syncing
    NetworkMessageType[NetworkMessageType["GetBlockHeadersRequest"] = 40] = "GetBlockHeadersRequest";
    NetworkMessageType[NetworkMessageType["GetBlockHeadersResponse"] = 41] = "GetBlockHeadersResponse";
    NetworkMessageType[NetworkMessageType["GetBlocksRequest"] = 42] = "GetBlocksRequest";
    NetworkMessageType[NetworkMessageType["GetBlocksResponse"] = 43] = "GetBlocksResponse";
    // Block gossip
    NetworkMessageType[NetworkMessageType["GetCompactBlockRequest"] = 60] = "GetCompactBlockRequest";
    NetworkMessageType[NetworkMessageType["GetCompactBlockResponse"] = 61] = "GetCompactBlockResponse";
    NetworkMessageType[NetworkMessageType["GetBlockTransactionsRequest"] = 62] = "GetBlockTransactionsRequest";
    NetworkMessageType[NetworkMessageType["GetBlockTransactionsResponse"] = 63] = "GetBlockTransactionsResponse";
    NetworkMessageType[NetworkMessageType["NewBlockHashes"] = 64] = "NewBlockHashes";
    NetworkMessageType[NetworkMessageType["NewCompactBlock"] = 65] = "NewCompactBlock";
    // Transaction gossip
    NetworkMessageType[NetworkMessageType["PooledTransactionsRequest"] = 80] = "PooledTransactionsRequest";
    NetworkMessageType[NetworkMessageType["PooledTransactionsResponse"] = 81] = "PooledTransactionsResponse";
    NetworkMessageType[NetworkMessageType["NewPooledTransactionHashes"] = 82] = "NewPooledTransactionHashes";
    NetworkMessageType[NetworkMessageType["NewTransactions"] = 83] = "NewTransactions";
})(NetworkMessageType = exports.NetworkMessageType || (exports.NetworkMessageType = {}));
