// @ironfish/sdk/build/src/rpc/routes/wallet/mintAsset.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MintAssetResponseSchema = exports.MintAssetRequestSchema = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const rust_nodejs_1 = require("@ironfish/rust-nodejs");
const yup = __importStar(require("yup"));
const assert_1 = require("../../../assert");
const utils_1 = require("../../../utils");
const router_1 = require("../router");
const utils_2 = require("./utils");
exports.MintAssetRequestSchema = yup
    .object({
    account: yup.string().required(),
    fee: utils_1.YupUtils.currency({ min: 1n }).defined(),
    value: utils_1.YupUtils.currency({ min: 1n }).defined(),
    assetId: yup.string().optional(),
    expiration: yup.number().optional(),
    expirationDelta: yup.number().optional(),
    confirmations: yup.number().optional(),
    metadata: yup.string().optional().max(rust_nodejs_1.ASSET_METADATA_LENGTH),
    name: yup.string().optional().max(rust_nodejs_1.ASSET_NAME_LENGTH),
})
    .defined();
exports.MintAssetResponseSchema = yup
    .object({
    assetId: yup.string().required(),
    hash: yup.string().required(),
    name: yup.string().required(),
    value: yup.string().required(),
})
    .defined();
router_1.routes.register(`${router_1.ApiNamespace.wallet}/mintAsset`, exports.MintAssetRequestSchema, async (request, node) => {
    const account = utils_2.getAccount(node.wallet, request.data.account);
    const fee = utils_1.CurrencyUtils.decode(request.data.fee);
    const value = utils_1.CurrencyUtils.decode(request.data.value);
    const expirationDelta = request.data.expirationDelta ?? node.config.get('transactionExpirationDelta');
    let options;
    if (request.data.assetId) {
        options = {
            assetId: Buffer.from(request.data.assetId, 'hex'),
            expiration: request.data.expiration,
            fee,
            expirationDelta,
            value,
            confirmations: request.data.confirmations,
        };
    }
    else {
        assert_1.Assert.isNotUndefined(request.data.name, 'Must provide name or identifier to mint');
        const metadata = request.data.metadata ?? '';
        options = {
            expiration: request.data.expiration,
            fee,
            metadata: metadata,
            name: request.data.name,
            expirationDelta,
            value,
            confirmations: request.data.confirmations,
        };
    }
    const transaction = await node.wallet.mint(account, options);
    assert_1.Assert.isEqual(transaction.mints.length, 1);
    const mint = transaction.mints[0];
    request.end({
        assetId: mint.asset.id().toString('hex'),
        hash: transaction.hash().toString('hex'),
        name: mint.asset.name().toString('hex'),
        value: mint.value.toString(),
    });
});
