// @ironfish/sdk/build/src/rpc/routes/wallet/postTransaction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostTransactionResponseSchema = exports.PostTransactionRequestSchema = void 0;
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
const yup = __importStar(require("yup"));
const rawTransaction_1 = require("../../../primitives/rawTransaction");
const router_1 = require("../router");
const utils_1 = require("./utils");
exports.PostTransactionRequestSchema = yup
    .object({
    account: yup.string().trim(),
    transaction: yup.string().defined(),
    broadcast: yup.boolean().optional(),
})
    .defined();
exports.PostTransactionResponseSchema = yup
    .object({
    accepted: yup.bool().optional(),
    broadcasted: yup.bool().optional(),
    hash: yup.string().defined(),
    transaction: yup.string().defined(),
})
    .defined();
router_1.routes.register(`${router_1.ApiNamespace.wallet}/postTransaction`, exports.PostTransactionRequestSchema, async (request, node) => {
    const account = utils_1.getAccount(node.wallet, request.data.account);
    const bytes = Buffer.from(request.data.transaction, 'hex');
    const raw = rawTransaction_1.RawTransactionSerde.deserialize(bytes);
    const { accepted, broadcasted, transaction } = await node.wallet.post({
        transaction: raw,
        account,
        broadcast: request.data.broadcast,
    });
    const serialized = transaction.serialize();
    request.end({
        accepted,
        broadcasted,
        hash: transaction.hash().toString('hex'),
        transaction: serialized.toString('hex'),
    });
});
